# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/nsICanvasRenderingContextInternal.h
# Commit: bbfe23c61add
# Full Hash: bbfe23c61add07377daf80696fd6996c0463f5b6
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-06-09 09:21:34
# Regressor Bug: 1632249
# File Overlap Count: 3
# Description:
#   Bug 1632249 - Support compositing out-of-process WebGL. r=handyman,lsalzman,nical,geckoview-reviewers,agi,imanol
#   
#   * Majorly simplity CanvasRenderer
#   * Replace GLScreenBuffer with trivial GLSwapChain
#   * Use descriptor structs so that future SharedSurface changes aren't so painful
# ==============================================================================

diff -r 8fcfac450ecc -r bbfe23c61add dom/canvas/nsICanvasRenderingContextInternal.h
--- a/dom/canvas/nsICanvasRenderingContextInternal.h	Mon Jun 08 20:13:01 2020 +0000
+++ b/dom/canvas/nsICanvasRenderingContextInternal.h	Mon Jun 08 20:34:15 2020 +0000
@@ -6,6 +6,8 @@
 #ifndef nsICanvasRenderingContextInternal_h___
 #define nsICanvasRenderingContextInternal_h___
 
+#include <memory>
+
 #include "mozilla/gfx/2D.h"
 #include "nsISupports.h"
 #include "nsIInputStream.h"
@@ -29,6 +31,7 @@
 class nsIDocShell;
 
 namespace mozilla {
+class ClientWebGLContext;
 class PresShell;
 namespace layers {
 class CanvasLayer;
@@ -37,6 +40,7 @@
 class Layer;
 class LayerManager;
 class LayerTransactionChild;
+class PersistentBufferProvider;
 class WebRenderCanvasData;
 }  // namespace layers
 namespace gfx {
@@ -57,6 +61,10 @@
 
   NS_DECLARE_STATIC_IID_ACCESSOR(NS_ICANVASRENDERINGCONTEXTINTERNAL_IID)
 
+  nsICanvasRenderingContextInternal()
+      : mSharedPtrPtr(
+            std::make_shared<nsICanvasRenderingContextInternal*>(this)) {}
+
   void SetCanvasElement(mozilla::dom::HTMLCanvasElement* parentCanvas) {
     RemovePostRefreshObserver();
     mCanvasElement = parentCanvas;
@@ -129,6 +137,10 @@
   virtual already_AddRefed<mozilla::gfx::SourceSurface> GetSurfaceSnapshot(
       gfxAlphaType* out_alphaType = nullptr) = 0;
 
+  virtual RefPtr<mozilla::gfx::SourceSurface> GetFrontBufferSnapshot() {
+    return GetSurfaceSnapshot();
+  }
+
   // If this is called with true, the backing store of the canvas should
   // be created as opaque; all compositing operators should assume the
   // dst alpha is always 1.0.  If this is never called, the context's
@@ -202,6 +214,13 @@
     return false;
   }
 
+  virtual void OnBeforePaintTransaction() {}
+  virtual void OnDidPaintTransaction() {}
+  virtual mozilla::layers::PersistentBufferProvider* GetBufferProvider() {
+    return nullptr;
+  }
+  virtual mozilla::ClientWebGLContext* AsWebgl() { return nullptr; }
+
   //
   // shmem support
   //
@@ -216,6 +235,9 @@
   RefPtr<mozilla::dom::HTMLCanvasElement> mCanvasElement;
   RefPtr<mozilla::dom::OffscreenCanvas> mOffscreenCanvas;
   RefPtr<nsRefreshDriver> mRefreshDriver;
+
+ public:
+  const std::shared_ptr<nsICanvasRenderingContextInternal* const> mSharedPtrPtr;
 };
 
 NS_DEFINE_STATIC_IID_ACCESSOR(nsICanvasRenderingContextInternal,